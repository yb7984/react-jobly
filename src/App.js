import NavigationBar from './NavigationBar';
import Home from './Home';
import Companies from './Companies';
import Company from './Company';
import Jobs from './Jobs';
import LoginForm from './LoginForm';
import SignUpForm from './SignUpForm';
import ProfileForm from './ProfileForm';
import { Switch, Route, useHistory } from 'react-router-dom';

import { library } from '@fortawesome/fontawesome-svg-core';
// import { fab } from '@fortawesome/free-brands-svg-icons';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';

import 'bootstrap/dist/css/bootstrap.min.css';
import { useState } from 'react';
import UserContext from './context/userContext';
import JoblyApi from './api';
import Loading from './Loading';

//add the using fontawesome
library.add(faSpinner);

/**
 * Here is where the app starts
 * @returns 
 */
function App() {
  const [loginUser, setLoginUser] = useState(null);
  const history = useHistory();

  let isLoading = !loginUser && JoblyApi.token;

  //load login user information if available
  const setUser = async () => {
    if (isLoading && !loginUser && JoblyApi.token) {
      const user = await JoblyApi.getCurrentUser();

      if (user) {
        setLoginUser(user);
      } else {
        setLoginUser(null);
      }

      isLoading = false;
    }
  }

  setUser();


  // function for user login
  const login = async (username, password) => {

    const result = await JoblyApi.login(username, password);

    if (result === true) {
      setLoginUser(await JoblyApi.getCurrentUser());

      return true;
    }

    return false;
  }

  //function for user sign up
  const signup = async (user) => {


    const result = await JoblyApi.register(user);

    if (result === true) {
      setLoginUser(await JoblyApi.getCurrentUser());

      return true;
    }

    return false;
  }

  // function for user logout
  const logout = () => {
    JoblyApi.token = "";
    setLoginUser(null);

    history.push('/');
  }

  if (isLoading) {
    return (<Loading />);
  }

  return (
    <div className="App">
      <UserContext.Provider value={{ loginUser, setLoginUser }}>
        <NavigationBar logout={logout} />
        <Switch>
          <Route path="/companies/:handle">
            <Company />
          </Route>
          <Route path="/companies">
            <Companies />
          </Route>
          <Route path="/jobs">
            <Jobs />
          </Route>
          <Route path="/profile">
            <ProfileForm />
          </Route>
          <Route path="/signup">
            <SignUpForm signup={signup} />
          </Route>
          <Route path="/login">
            <LoginForm login={login} />
          </Route>
          <Route exact path="/">
            <Home />
          </Route>
          <Route>
            404
        </Route>
        </Switch>
      </UserContext.Provider>
    </div>
  );
}

export default App;
