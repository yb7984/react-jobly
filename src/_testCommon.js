"use strict";

//mock localStorage
const localStorageMock = {
  data: {},
  getItem: (key) => {
    return data[key] === undefined ? null : data[key];
  },
  setItem: (key, value) => {
    data[key] = value;
  }
}

// token ("testuser" / "password")
const testToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZ" +
  "SI6InRlc3R1c2VyIiwiaXNBZG1pbiI6ZmFsc2UsImlhdCI6MTU5ODE1OTI1OX0." +
  "FtrMwBQwe6Ue-glIFgz_Nf8XxRT2YecFCiSpYL0fCXc";

module.exports = {
  localStorageMock,
  testToken
};
