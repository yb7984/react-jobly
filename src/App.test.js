import { fireEvent, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import App from './App';
import axiosMock from 'axios';
import { localStorageMock, testToken } from './_testCommon';
import JoblyApi from './api';

console.error = jest.fn((msg) => {
  console.log(msg);
});

beforeEach(() => {
});

afterEach(() => {
  JoblyApi.token = "";
});

test('renders page without login', () => {
  JoblyApi.token = "";

  render(<MemoryRouter initialEntries={['/']}><App /></MemoryRouter>);
  render(<MemoryRouter initialEntries={['/login']}><App /></MemoryRouter>);
  render(<MemoryRouter initialEntries={['/signup']}><App /></MemoryRouter>);
});


test('match a snapshot, render with login', async () => {

  JoblyApi.token = testToken;

  const data = {
    status: 200,
    data: {
      user: {
        username: "testuser",
        firstName: "test_f",
        lastName: "test_l",
        isAdmin: true,
        applications: []
      }
    },
  };

  axiosMock.mockResolvedValueOnce(data);
  axiosMock.mockRejectedValue({
    data: {
      error: {
        message: "error"
      }
    }
  })

  const { asFragment, getByTestId, getByText, findByText, findByPlaceholderText } = render(<MemoryRouter initialEntries={['/']}><App /></MemoryRouter>);

  expect(asFragment()).toMatchSnapshot();

  expect(getByTestId("loading")).toBeInTheDocument();

  await findByText('All the jobs in one convenient place.');

  expect(getByText('All the jobs in one convenient place.')).toBeInTheDocument();


  axiosMock.mockResolvedValueOnce({
    status: 200,
    data: {
      companies: [
        {
          handle: "test",
          name: "Test Company",
          numEmployees: 200,
          description: "test test"
        }
      ]
    }
  });

  // go to companies route
  fireEvent.click(getByText("Companies"));

  await findByPlaceholderText('Search companies here');

  expect(getByText("Test Company")).toBeInTheDocument();


  axiosMock.mockResolvedValueOnce({
    status: 200,
    data: {
      jobs: [
        {
          id: "test",
          title: "Test Jobs",
          salary: 200000,
          equity: 70,
          companyHandle: "testcom",
          companyName: "Test Company"
        }
      ]
    }
  });

  // go to jobs route
  fireEvent.click(getByText("Jobs"));

  await findByPlaceholderText('Search jobs here');

  expect(getByText("Test Company")).toBeInTheDocument();
});
