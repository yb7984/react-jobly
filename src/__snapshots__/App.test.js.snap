// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match a snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="navbar navbar-expand-sm navbar-light bg-light"
    >
      <div
        class="container"
      >
        <span
          class="navbar-brand"
        >
          <a
            aria-current="page"
            class="fw-bolder text-primary active"
            href="/"
            style="font-weight: bold; color: black; text-decoration: none;"
          >
            Jobly
          </a>
        </span>
        <button
          aria-controls="basic-navbar-nav"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="justify-content-end navbar-collapse collapse"
        >
          <div
            class="me-auto navbar-nav"
          />
        </div>
      </div>
    </nav>
    <div
      class="Home d-flex flex-column justify-content-center text-center"
    >
      <div
        class="display-2 p-2"
      >
        Jobly
      </div>
      <div
        class="display-5 p-2"
      >
        All the jobs in one convenient place.
      </div>
      <div
        class="p-2"
      >
        <a
          class="btn btn-primary btn-lg mx-2"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-primary btn-lg mx-2"
          href="/signup"
        >
          Singup
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`match a snapshot, render with login 1`] = `
<DocumentFragment>
  <div
    class="Loading display-1 d-flex flex-column justify-content-center text-center"
    data-testid="loading"
  >
    <div>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-spinner fa-w-16 fa-spin fa-lg "
        data-icon="spinner"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
